### ISSUE TITLE
AES CTR error

### ISSUE THREAD

**Issue by sticfeynman** [OP-TEE/optee_os#2305]
Hi,
in the Aarch32 accelerated crypto environment,  I test AES-CTR with a pattern:

`static uint8_t ctr_in[] = {
0x01, 0x01, 0x01, 0x01, 0x01, 0x01, 0x01, 0x01, 0x01, 0x01, 0x01, 0x01, 0x01, 0x01, 0x01, 0x01,
0x01, 0x01, 0x01, 0x01, 0x01, 0x01, 0x01, 0x01, 0x01, 0x01, 0x01, 0x01, 0x01, 0x01, 0x01, 0x01,
0x01, 0x01, 0x01, 0x01, 0x01, 0x01, 0x01, 0x01, 0x01, 0x01, 0x01, 0x01, 0x01, 0x01, 0x01, 0x01,
0x01, 0x01, 0x01, 0x01, 0x01, 0x01, 0x01, 0x01, 0x01, 0x01, 0x01, 0x01, 0x01, 0x01, 0x01, 0x01,
0x01, 0x01, 0x01
};`
`
static uint8_t ctr_out[] = {
0xbb, 0xfe, 0x07, 0x04, 0x1c, 0x8e, 0x09, 0x61, 0xfb, 0xb1, 0x7c, 0xa5, 0x4d, 0x2b, 0x30, 0xf6,
0x26, 0x9e, 0xff, 0x61, 0x18, 0x47, 0xc6, 0x06, 0x81, 0x02, 0x84, 0xcd, 0x9c, 0x4b, 0x6d, 0x21,
0xe2, 0x64, 0xa6, 0x50, 0x7f, 0x28, 0x81, 0x6f, 0x29, 0xda, 0xd5, 0x56, 0x3f, 0x46, 0xac, 0xca,
0x37, 0xe7, 0x77, 0x36, 0xbc, 0x76, 0x39, 0x57, 0xaa, 0x67, 0x1b, 0x2a, 0xe6, 0x36, 0x57, 0x6d,
0x2a, 0xb8, 0x77
};`
`
uint8_t ctr_key[] = {
0x2, 0x2, 0x2, 0x2, 0x2, 0x2, 0x2, 0x2, 0x2, 0x2, 0x2, 0x2, 0x2, 0x2, 0x2, 0x2
};`
`
uint8_t ctr_iv[] = {
0x3, 0x3, 0x3, 0x3, 0x3, 0x3, 0x3, 0x3, 0x3, 0x3, 0x3, 0x3, 0x3, 0x3, 0x3, 0x3
};`

`
XTEST_CIPH_CASE(TEE_ALG_AES_CTR, TEE_TYPE_AES,
			ctr_key, ctr_iv, 11,
			ctr_in,
			ctr_out)`

And I got such errors:

Got
  BB:FE:07:04 1C:8E:09:61 FB:B1:7C:A5 4D:2B:30:F6
  26:9E:FF:61 18:47:C6:06 81:02:84:CD 9C:4B:6D:21
  E2:64:A6:50 7F:28:81:6F 29:DA:D5:56 3F:46:AC:CA
  37:E7:77:36 BC:76:39:57 AA:67:1B:2A E6:36:57:6D
  3E:68:B2
Expected
  BB:FE:07:04 1C:8E:09:61 FB:B1:7C:A5 4D:2B:30:F6
  26:9E:FF:61 18:47:C6:06 81:02:84:CD 9C:4B:6D:21
  E2:64:A6:50 7F:28:81:6F 29:DA:D5:56 3F:46:AC:CA
  37:E7:77:36 BC:76:39:57 AA:67:1B:2A E6:36:57:6D
  2A:B8:77

Got
  01:01:01:01 01:01:01:01 01:01:01:01 01:01:01:01
  01:01:01:01 01:01:01:01 01:01:01:01 01:01:01:01
  01:01:01:01 01:01:01:01 01:01:01:01 01:01:01:01
  01:01:01:01 01:01:01:01 01:01:01:01 01:01:01:01
  15:D1:C4 
Expected
  01:01:01:01 01:01:01:01 01:01:01:01 01:01:01:01
  01:01:01:01 01:01:01:01 01:01:01:01 01:01:01:01
  01:01:01:01 01:01:01:01 01:01:01:01 01:01:01:01
  01:01:01:01 01:01:01:01 01:01:01:01 01:01:01:01
  01:01:01

**Comment by jenswi-linaro** [comment]
This looks like an old problem, which version are you running?

**Comment by sticfeynman** [comment]
I'm running on 2.4.0 with commit e39aae81e1a40ba495893f1c4e04b23401eca3a3 ("core: crypto: arm32: add counter increment in ce_aes_ctr_encrypt()").

**Comment by jenswi-linaro** [comment]
This is a know bug in one of the older releases. It's fixed in `3.1.0` by
e39aae81e1a4 ("core: crypto: arm32: add counter increment in ce_aes_ctr_encrypt()")

**Comment by jenswi-linaro** [comment]
Sorry, missed that you already where using that commit.

**Comment by sticfeynman** [comment]
In my test case, the input are divided into three rounds and they are 16 bytes, 48 bytes and 3 bytes.
I observe that for this 48 bytes round the counter starts from 3 before entering `accel_ctr_encrypt()` and it becomes 7 when accel_ctr_encrypt() finishes. This seems wrong.

I noticed that there is one ctr addition in `aes_ctr_encrypt_nblocks()` and three addition in the .Lctrloop3x section of ce_aes_ctr_encrypt().  It seems not reasonable.

I did some experiment to avoid falling through .Lctrloop3x section, i.e. I replaced `bcs		.Lctrloop` with `b .Lctrloop` and I passed the test.
